### DISCUSSION TITLE
Nengo with 5D data

### DISCUSSION THREAD

**Discussion by cnrgbot** [nengo/nengo-zoo#895] [General Discussion]
> _Archived from the Nengo Discourse forum._ Originally posted by **ji16qewe** on 2021-03-17. (Forum topic #1654, post #1 / id 4921)

Dear all,

I am converting an analog signal through an ADC, and my output is actually 5D arrays.
Specifically, it has the shape of (1438, 95, 3, 32,32). I need to iterate over the first dimension and send them into a simple spiking network batch by batch. However, I am not sure if Nengo core is able to deal with 4-d data?

My main aim would be to create an end-to-end spiking neural network (the inserted data should be changed into spikes, and learning should be done in a supervised way) that can do classifications based on the inserted 5D-data (4D after sending batches of it). However, I am not sure if I should use Nengo core or Nengo DL for that.

Also, a very simple code that is capable of doing that in Nengo core or Nengo DL would be quite helpful. As I am still new to Nengo.

Thanks.

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **zerone** on 2021-03-17. (Forum topic #1654, post #2 / id 4924)

Hello @​ji16qewe, welcome to our community. For the following question, 

> *Originally @​ji16qewe wrote [post #1, topic #1654]:*
>
> However, I am not sure if Nengo core is able to deal with 4-d data?

I am pointing you towards two Nengo Core tutorials which show you how to [represent `N` dimensional data](https://www.nengo.ai/nengo/examples/basic/2d-representation.html) and do [calculations on them](https://www.nengo.ai/nengo/examples/basic/multiplication.html). Do note that in these tutorials the value of `N=2`, but the principle of representing `4D` data would remain the same.

Basically, you create `N` one dimensional ensembles, e.g. 

`A = nengo.Ensemble(100, dimensions=1, radius=10)` and 
`B = nengo.Ensemble(100, dimensions=1, radius=10)` 

(for the value of `N=2`) and then create another ensemble for representing `N=2` dimensional data, i.e. 

`combined = nengo.Ensemble(220, dimensions=2, radius=15)`

And then you need to connect them as follows:

```
nengo.Connection(A, combined[0])
nengo.Connection(B, combined[1])
```
You will find more details in the linked tutorials. 

Next, with respect to the following:
> *Originally @​ji16qewe wrote [post #1, topic #1654]:*
>
> My main aim would be to create an end-to-end spiking neural network (the inserted data should be changed into spikes, and learning should be done in a supervised way) that can do classifications based on the inserted 5D-data (4D after sending batches of it).
I am not completely sure if Nengo or Nengo-DL allows spike based training as of now. There's `KerasSpiking` which allows you to do [spike aware training](https://www.nengo.ai/keras-spiking/examples/spiking-fashion-mnist.html#Spiking-aware-training). Specifically, it does spike based forward pass and non-spike based backward pass (to propagate the errors back) during training phase. 

With Nengo-DL you can have an end-to-end spiking network - with room for some exceptions like MaxPooling op, unsupported neuron types e.g. softmax etc. which do not have a spiking counterpart. How you can do training and inference in Nengo-DL is as follows. You will first train your network with non-spiking neurons (e.g. ReLU) and then convert your model to spiking one by using `nengo_dl.Converter()` API and then proceed with inference. An excellent tutorial on training and inference with Nengo-DL is [here](https://www.nengo.ai/nengo-dl/examples/keras-to-snn.html).

With respect to following about training and inference with 4D data:
> *Originally @​ji16qewe wrote [post #1, topic #1654]:*
>
> Also, a very simple code that is capable of doing that in Nengo core or Nengo DL would be quite helpful.

you will find how it accepts a 3D data in the above linked Nengo-DL tutorial. Basically, you create your usual TF network architecture to accept the `N` dimensional data and then while working in Nengo-DL context, you flatten your data before passing it to the Nengo-DL simulator functions. 

If your task is to do deep learning, I would suggest you to look at [Nengo-DL](https://www.nengo.ai/nengo-dl/index.html) and [KerasSpiking](https://www.nengo.ai/keras-spiking/), evaluating both which suits you best.

**Reply by cnrgbot** [reply]
> _Archived from the Nengo Discourse forum._ Originally posted by **ji16qewe** on 2021-03-17. (Forum topic #1654, post #3 / id 4927)

@​zerone  First of all, thanks a lot for your reply.

However, what I got from your comment that Nengo so far does not support end-to-end training in the spiking domain. Because I see that either KerasSpiking or Nengo DL is trained normally in the non-spiking domain, then we just convert the activation functions into spiking version in the test phase, am I right?

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **zerone** on 2021-03-17. (Forum topic #1654, post #4 / id 4928)

Nengo supports unsupervised PES training rule in spiking domain, this I know of. About supporting state-of-the-art backprop methods for spike based training, I have not seen any example so far. Although, Nengo developers would answer it more confidently. 

About KerasSpiking and Nengo-DL, your observation is correct.

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **xchoo** on 2021-03-20. (Forum topic #1654, post #5 / id 4935)

> *Originally @​ji16qewe wrote [post #3, topic #1654]:*
>
> However, what I got from your comment that Nengo so far does not support end-to-end training in the spiking domain.

Currently, Nengo only supports end-to-end training in the spiking domain for [online learning](https://www.nengo.ai/nengo/examples.html#learning). For offline and batch learning, you'll need to use something like the KerasSpiking package for TensorFlow.

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **zerone** on 2021-03-20. (Forum topic #1654, post #6 / id 4937)

Hello @​xchoo, I am new to online learning. Is the Nengo supported online learning supervised or unsupervised?

**Reply by cnrgbot** [reply]
> _Archived from the Nengo Discourse forum._ Originally posted by **xchoo** on 2021-03-23. (Forum topic #1654, post #7 / id 4963)

Nengo supports both [supervised](https://www.nengo.ai/nengo/examples/learning/learn-communication-channel.html) and [unsupervised](https://www.nengo.ai/nengo/examples/learning/learn-unsupervised.html) online learning. You can build your networks with the different forms of learning by using different learning rules. Included in Nengo is the [PES](https://www.nengo.ai/nengo/frontend-api.html#nengo.PES) learning rule, which is a online supervised learning rule, as well as the [BCM](https://www.nengo.ai/nengo/frontend-api.html#nengo.BCM) and [Oja](https://www.nengo.ai/nengo/frontend-api.html#nengo.Oja) learning rules, which are online unsupervised learning rules. Nengo also includes other learning rules, which are described [here](https://www.nengo.ai/nengo/frontend-api.html#module-nengo.learning_rules). There is even an (super rough) STDP implementation described [here](https://forum.nengo.ai/t/nengo-wta-and-homeostasis/1439/10) and [here](https://forum.nengo.ai/t/unsupervised-learning-of-heteroassociations/1505/2) (STDP is an unsupervised learning rule).

The cool thing about Nengo is that it is general enough to users to implement their own custom learning rule, be they supervised or not! :smiley:

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **zerone** on 2021-03-23. (Forum topic #1654, post #8 / id 4967)

Thanks @​xchoo for these sources. If one looks to dive into online learning with spiking neurons, other than these sources, do you suggest something more basic to begin with?

**Reply by cnrgbot** [reply]
> _Archived from the Nengo Discourse forum._ Originally posted by **xchoo** on 2021-03-23. (Forum topic #1654, post #9 / id 4968)

> *Originally @​zerone wrote [post #8, topic #1654]:*
>
> If one looks to dive into online learning with spiking neurons, other than these sources, do you suggest something more basic to begin with?

Apart from the examples on the Nengo documentation, there is @​tbekolay's Master thesis which goes into some detail of the mathematics behind online learning with spiking neurons. I referenced his thesis in this [forum post](https://forum.nengo.ai/t/is-the-pes-learning-algorithm-biologically-interpretable/1658/2). 

@​tcstewar has some online learning examples designed for Nengo available [here](https://github.com/tcstewar/nengo_learning_examples) as well that might be useful to you. Although, some of those example may have to be updated for the latest version of Nengo.

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **zerone** on 2021-03-23. (Forum topic #1654, post #10 / id 4970)

Awesome! Thanks!

**Reply by cnrgbot** [reply]
> _Archived from the Nengo Discourse forum._ Originally posted by **xchoo** on 2021-03-23. (Forum topic #1654, post #11 / id 4972)

Just an update, @​tcstewar has also referred me to this [Jupyter notebook](https://github.com/tcstewar/syde556/blob/master/SYDE%20556%20Lecture%2010%20Learning.ipynb) which may be helpful as a starting guide for online learning in Nengo.

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **zerone** on 2021-03-23. (Forum topic #1654, post #12 / id 4974)

This looks like a great source! Thanks!
